Shared Resources via OpenShift Builds Operator (#1457)
openshift/enhancements: enhancements/cluster-scope-secret-volumes/shared-resources-via-openshift-builds-operator.md

  - https://issues.redhat.com/browse/RHDP-701
see-also:
  - "enhancements/cluster-scope-secret-volumes/csi-driver-host-injections.md"
  - "enhancements/cluster-scope-secret-volumes/shared-resource-validation.md"
replaces:
  - "/enhancements/subscription-content/subscription-injection.md"
superseded-by: []
---

# Shared Resources via OpenShift Builds Operator

## Summary

Remove the Shared Resources CSI driver from OpenShift, and deliver it to customers with the
forthcoming OpenShift Builds OLM Operator.

## Motivation

The Shared Resources CSI driver provides a novel way for sharing information across OpenShift
namespaces. Though there was hope this capability would be generally useful, in practice this
feature is mainly needed for Red Hat build workloads.

Packaging this driver with the forthcoming OpenShift Builds operator provides several advantages:

- Provides an extension to OpenShift Builds, which is based on the upstream [Shipwright](https://shipwright.io) project.
- Allows release outside of OpenShift's cadence.
- Potential to fully support customers on older versions of OCP.

> - Pathway to adoption on hosted control plane clusters.

Reviewer: also add something around it NOT being part of the base OCP payload, and that optional by default lends well to https://issues.redhat.com/browse/OCPSTRAT-421

set of feature sets will need to be updated (ex - openshift/library-go).

The Shared Resource CSI driver itself will need to be removed from the OpenShift payload. Its
components will need to be re-productized as an OLM operand.

### Risks and Mitigations

**Risk**: Same CSI driver will be deployed twice on tech preview clusters.
*Mitigation*: Operator will check if the `CSIDriverSharedResource` feature is enabled before
deploying the Shared Resource CSI driver itself.

**Risk**: Admins or users will want the Shared Resource CSI Driver without Shipwright
*Mitigation*: Shipwright components can be disabled via the `OpenshiftBuild` CRD.

### Drawbacks

- Shared resources no longer a part of OpenShift
- Customers who need entitlements in builds will need to install a new operator, or resort to less
  desirable work-arounds
- Potential impact on Red Hat components that need shared resources and/or RHEL entitlements

## Design Details

### Test Plan

- Shared resource logic in the OCP Builds tech preview suite will need to be migrated to the
  OpenShift Builds operator test suites
- OpenShift Builds operator will need a techpreview variant to ensure the shared resource CSI

>   driver is not managed by the operator on OCP 4.15 and earlier (tech preview).

Reviewer 2: Can you add a line about moving (or removing) the e2e tests in openshift/origin that depend on the shared resource driver?
<https://github.com/openshift/origin/blob/master/test/extended/builds/volumes.go>
<https://github.com/openshift/origin/blob/master/test/extended/storage/inline.go>

Author: Updated

  - "@bparees"
api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
  - "@bparees"
creation-date: 2023-08-17
last-updated: 2023-08-22
tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
  - https://issues.redhat.com/browse/RHDP-701
see-also:
  - "enhancements/cluster-scope-secret-volumes/csi-driver-host-injections.md"
  - "enhancements/cluster-scope-secret-volumes/shared-resource-validation.md"
replaces:
  - "/enhancements/subscription-content/subscription-injection.md"
superseded-by: []
---

# Shared Resources via OpenShift Builds Operator

## Summary

Remove the Shared Resources CSI driver from OpenShift, and deliver it to customers with the
forthcoming OpenShift Builds OLM Operator.

## Motivation

The Shared Resources CSI driver provides a novel way for sharing information across OpenShift
namespaces. Though there was hope this capability would be generally useful, in practice this

> feature is mainly needed for Red Hat build workloads.

Reviewer: we've asked mgmt/staffing engineers quite a number of times since its tech preview introduction back in 4.7 about actually plans for non build use cases and have been told "no plans" and that this thing is really only a "build" thing.  

Also, according to PMs, all the 20 some odd external customers listed in OCPSTRAT-201 are characterized as only interested for builds.

So given finite resource constraints, we are admittedly focused on sorting out the majority use case vs. others which have been suggested, but never got prioritized.  We are past the point of accommodating "might be nice". The time betwen 4.7 and 4.14 seems like a sufficient runway to see if these non-build use cases would get prioritized.

The harsh reality as well is that with the old "developer experience" team no longer part of core OCP for a couple of years now,  but still the owner of shared resources and builds (where attempts to transfer shared resource ownership to within OCP have been rejected), with each passing release cycle, we are more and more out of touch with OCP processes, schedules, fine grained technical details, among others.  Shipping shared resources in the same way we ship all our other deliverables bypasses a lot of "indirect" costs outside of the actual coding/testing/etc.

Packaging this driver with the forthcoming OpenShift Builds operator provides several advantages:

- Provides an extension to OpenShift Builds, which is based on the upstream [Shipwright](https://shipwright.io) project.
- Allows release outside of OpenShift's cadence.
- Potential to fully support customers on older versions of OCP.

> - Pathway to adoption on hosted control plane clusters.

Reviewer 3: This seems to be answered in Alternatives -> Remain in OCP section, I think it would be useful to elaborate this issue in the motivation (is it a main motivation of this change actually?)

> - Pathway to adoption on hosted control plane clusters.

Reviewer: I would agree @petr-muller.   This was a key motivation in fact. WDYT @adambkaplan ?

> - Pathway to adoption on hosted control plane clusters.

Author: Updated motivation

### User Stories

- As a developer installing RHEL RPMs in my container image, I want to use the Shared Resource CSI
  driver to mount my cluster's entitlement into my build.
- As a cluster admin of a build/CI cluster, I want to manage the Shared Resource CSI driver via an

>   operator so that its installation, custom resources, and lifecycle are managed by OLM.

Reviewer 4: more work than it being in the payload?  yes.  But in general we are trying to break apart out monolith.

In the future things like "platform operators" may make this management of add-on operators easier/better.

>   operator so that its installation, custom resources, and lifecycle are managed by OLM.

Reviewer: Yep my comment https://github.com/openshift/enhancements/pull/1457/files#r1301768942 was my attempt to note what @bparees is referring to, and that becoming "optional by default" is a path that better lines up with how developer org provides function on top of OCP.

>   operator so that its installation, custom resources, and lifecycle are managed by OLM.

Author: > Sounds like more work for me as an admin?

Not as much as you think, and provisioning can be automated via GitOps. We do this with RHTAP today for OpenShift Pipelines and OpenShift GitOps itself.

The OpenShift Builds operator will be enhanced to support the lifecycle and management of the
Shared Resource CSI Driver and its associated components. This will be in addition to the
lifecycle management of Shipwright, whose logic is implemented in the upstream
[Shipwright Operator](https://github.com/shipwright-io/operator).

### Workflow Description

1. Cluster admin installs the OpenShift Builds operator via OLM, using one of the following:
   1. OperatorHub web console in OpenShift
   2. Creating an appropriate OLM `Subscription` object via `oc apply`, OpenShift GitOps, and so
      forth.
2. Once the operator is deployed, it creates an `OpenShiftBuild` custom resource instance on the
   cluster:

   ```yaml
   apiVersion: operator.build.openshift.io/v1alpha1
   kind: OpenShiftBuild
   metadata:
     name: cluster
   spec:
     shipwright:
       build:
         enabled: false
     sharedResources:
       enabled: false
   ```

3. The OpenShift Builds Operator reconciles the `OpenShiftBuild` instance (singleton), and reports
   an appropriate status. If `spec.sharedResources.enabled` is `true`, the operator deploys the

>    Shared Resource CSI driver, custom resource definitions, and webhook.

Reviewer 4: what happens when this is switched back to "false" after having been set to true?

Reviewer: My immediate thought to answering @bparees 's question is that "shared resources is uinstalled"

But then how much do we worry about, or can we even properly, "fail gracefully" for BuildConfig's that remain that point to the shared resource CSI driver for one of its volumes?  Does feel like perhaps the EP has to lay out the plan for that.

Does it somehow update any BuildConfigs using shared resources CSI volumes?

Author: @jsafrane may have to weigh in on the "right" way to remove a CSI driver.

the respective OCP feature gate enabled. This generally applies to tech preview clusters or
clusters with custom fature gate tunings.

With the start of OCP 4.16, the Shared Resource CSI driver operator will be removed from the
[Cluster Storage Operator](https://github.com/openshift/cluster-storage-operator). The feature
will likewise be removed from OpenShift's feature sets - any code that needs the most up to date
set of feature sets will need to be updated (ex - openshift/library-go).
The Shared Resource CSI driver itself will need to be removed from the OpenShift payload. Its
components will need to be re-productized as an OLM operand.

Notably, the design of this operator does **_not_** intend to use OLM API dependency resolution
to automate the deployment and management of OpenShift Pipelines. Cluster admimins will need to
separately (and manually) install OpenShift Pipelines.

### Risks and Mitigations

#### Duplicate CSI Drivers

On a tech preview cluster, the Same CSI driver could be deployed twice on tech preview clusters. The operator will check if the `CSIDriverSharedResource` feature is enabled before

> deploying the Shared Resource CSI driver itself.

Reviewer 4: would it be better/safer to check for the presence of the CSI driver itself, rather than indirectly making the determination based on feature gates?

Reviewer 4: also what happens if someone installs this operator/enables the csi driver and *then* turns on the techpreview featureset?

Reviewer: can the new build operator "disable" the `CSIDriverSharedResource` feature gate that is included in the the tech preview one per https://github.com/openshift/api/blob/master/config/v1/types_feature.go#L169 *after* tech preview is turned on?

otherwise, I wonder it simply renaming the shared resource CSI driver to something different than what the tech preview installs would suffice here.  Then, any OCP builds that define CSI volumes would point to the new driver and that would bypass the tech perview version of the driver.

WDYT @adambkaplan @bparees ?

I'm also good with adding a CSI drvier instance check on top of the feature gate check.

WDYT @adambkaplan ?

Reviewer 4: > can the new build operator "disable" the CSIDriverSharedResource feature gate that is included in the the tech preview one per https://github.com/openshift/api/blob/master/config/v1/types_feature.go#L169 after tech preview is turned on?

i don't think so, the a featureset (such as "techpreviewnoupgrade") is basically a static set of featuregates.  you can't turn off a specific gate, or remove it from the featureset (without shipping a new ocp release)

> otherwise, I wonder it simply renaming the shared resource CSI driver to something different than what the tech preview installs would suffice here. 

that could work, i don't know what impact is has on the buildconfig integration work that was done though.  might need to be some new/rework done there if it was renamed.

Reviewer: ACK on the feature gate.

On the integration, the *corev1.CSIVolumeSource type is used directly in the build type per https://github.com/openshift/api/blob/master/build/v1/types.go#L1457C6-L1457C29 so it is up the user to specify the driver name in the BuildConfig.  There are no code changes in the OCP/OCM build controller to switch driver names.

Reviewer 4: ah, ok, so it was generic support for CSI volumes that was added, not something specific to shared resource volumes.  cool

Author: I like using a different name for registering the CSI Driver, for two reasons:

1. Avoiding the driver name and deployment collision with OCP.
2. Branding - the current driver name has the `openshift.io` domain/api group.  Using a different api group (ex: `sharedresources.dev`) will make the core idea more upstream-friendly.

#### Unwanted components/CRDs

Admins or users may want to use the Shared Resource CSI Driver without Shipwright or Tekton - for
example, developers or teams that heavily rely on the `BuildConfig` APIs today. To address this
concern, the OpenShift Builds operator will not rely on OLM API resolution to automate the
installation and deployment of OpenShift Pipelines. The operator will instead report an appropriate
status message if `spec.shipwright.build.enabled` is set to `true` and it detects that OpenShift
Pipelines has not been installed.

### Drawbacks

- Shared resources no longer a part of OpenShift.
- Customers who need entitlements in builds will need to install a new operator, or resort to
  current cumbersome work-arounds.
- Potential impact on Red Hat components that need shared resources and/or RHEL entitlements.
- Manual installation of OpenShift Pipelines is required to deploy Shipwright Builds. Shipwright's

>   upstream operator currently installs Tekton via [OLM API dependency resolution](https://olm.operatorframework.io/docs/concepts/olm-architecture/dependency-resolution/).

Reviewer 4: to be honest i think avoiding the specifying of olm dependencies is a safer choice right now anyway.  That may(hopefully will) change with olm v1, but olm v0 dependencies have significant challenges.

  This proposal will deviate from the "upstream" behavior to support customers who want the Shared
  Resource CSI Driver, but are not interested in Shipwright or Tekton.

## Design Details

### Test Plan

- Shared resource logic in the OCP Builds tech preview suite will need to be migrated to the
  OpenShift Builds operator test suites

> - OpenShift Builds operator will need a techpreview variant to ensure the shared resource CSI

Reviewer 4: i'm not sure what this means or why it would be needed? 

maybe you mean "needs a test that runs in the techpreview variant to....."?

i'll reiterate here that i think it might be better if the operator directly looked for evidence that the CSI driver was already installed/enabled.  (Speaking of, i assume this is a cluster scoped operator only? because if not we need to talk about what happens if it gets installed multiple times, also)

and still need coverage for what happens when "builds operator is installed + shared resources enabled *before* someone turns on techpreview on the cluster"

Reviewer: I think different names for the CSI driver obj itself and a new namepace for the daemonset / webhooks would be sufficient for making sure the new operator only manages the GA version, and we leave the tech preview version to the OCP cluster storage operator based stuff 

Or do you all think I am missing something @bparees @adambkaplan

Reviewer 4: i generally like that approach, again keeping in mind that the existing buildconfig integrations that support the TP csi driver might need updating/revising.

Reviewer 4: ack, thanks

Author: Updated with new direction on CSI driver naming

Replacing the current validating webhook with
[CRD CEL expressions](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules)
(available in OCP 4.12) would prove to be challenging for the following reasons:

- The webhook validates pod admission to check that `readOnly: true` is set on the referenced
  volume. This is mainly to improve user experience, and could be removed due to scalability
  concerns.
- The webhook ensures that Secrets and ConfigMaps in any `openshift-*` namespace are not shared,
  with the exception of those explicitly allowed by the driver webhook's
  [configuration](https://github.com/openshift/csi-driver-shared-resource-operator/blob/master/assets/webhook/deployment.yaml#L37).

### Deploy Shared Resources as a Separate Operator

Rather than bundle the Shared Resources CSI driver with OpenShift Builds, the driver could be
deployed with its own OLM operator. The primary issue with this approach is that OpenShift Builds
would still want to deploy and manage this driver, since this is part of our strategy to make use

> of entiltements in builds simpler. This would likely add a lot of unnecessary complexity:

Reviewer 4: ```suggestion
of entitlements in builds simpler. This would likely add a lot of unnecessary complexity:
```

- An additional operator CRD to reconcile (ex: `SharedResourceConfig`)

> - Version skews when OLM does its API resolution, OR

Reviewer 4: version skews between what and what?

Author: Updated to reflect the concern with OLM dependency resolution

   spec:
     shipwright:
       build:
         managementState: Enabled
     sharedResources:
       managementState: Enabled
   ```

3. Each component has a `managementState` field that indicates if it should be disabled, enabled,
   or removed. These states implement the following state machine:
   1. `Disabled` components can only transition to the `Enabled` state.
   2. `Enabled` components can only transition to the `Removed` state.
   3. `Removed` components can only transition to the `Enabled` state.
3. The OpenShift Builds Operator reconciles the `OpenShiftBuild` instance (singleton), and reports
   an appropriate status. If `spec.sharedResources.managementState` is `Enabled`, the operator deploys the
   Shared Resource CSI driver, custom resource definitions, and webhook.
4. The operator will likewise deploy and manage Shipwright components using the `spec.shipwright.*`
   fields, respecting the `managementState: Enabled|Disabled|Removed` value for each.
   Reconciliation will reuse upstream Shipwright operator CRDs and controllers.
5. **The operator will enable Shipwright Builds and the Shared Resources CSI driver by default.**

### Variations

#### Shipwright Builds

> - To disable Shipwright Builds by default, set the `SHIPWRIGHT_BUILD_DEFAULT_DISABLED` envionrment

Author: missed typo.

```suggestion
- To disable Shipwright Builds by default, set the `SHIPWRIGHT_BUILD_DEFAULT_DISABLED` environment
```

  variable to `true` in the operator `Subscription` object:

  ```yaml
  kind: Subscription
  apiVersion: operators.coreos.com/v1alpha1
  metadata:
    name: openshift-build
    namespace: openshift-operators
  spec:
    ...
    config:
      env:
        - name: SHIPWRIGHT_BUILD_DEFAULT_DISABLED
          value: "true"
  ```

- If OpenShift Pipelines is not installed and `spec.shipwright.build.managementState` is set to
  `Enabled`, the operator does not install Shipwright Builds and reports that OpenShift Pipelines
  should be installed in its status.

> - If Shipwright Builds is enabled and then later disabled, the following should occur:

Author: Clarify in light of new management state API:

```suggestion
- If Shipwright Builds is enabled and then later set to `Removed`, the following should occur:
```

  - Shipwright build controllers should be deleted.
  - CRDs and associated admission webhooks should remain in place.

#### Shared Resource CSI Driver

- To disable the Shared Resource CSI Driver by default, set the `SHARED_RESOURCES_DEFAULT_DISABLED`
  environment variable to `true` in the operator `Subscription` object:

  ```yaml
  kind: Subscription
  apiVersion: operators.coreos.com/v1alpha1
  metadata:
    name: openshift-build
    namespace: openshift-operators
  spec:
    ...
    config:
      env:
        - name: SHARED_RESOURCES_DEFAULT_DISABLED
          value: "true"
  ```

> - If the CSI driver is disabled after it has been enabled, the following should occur:

Author: Fix to use the proper API state:

```suggestion
- If the CSI driver is `Removed` after it has been enabled, the following should occur:
```

  - The `CSIDriver` object for the driver should be deleted.
  - The `DaemonSet` for the CSI driver should be deleted.
  - CRDs and associated admission webhooks should remain in place.

### API Extensions

> The OpenShift Builds will introduce a new configuration CRD, `OpenShiftBuild`. It will contain 

Author: Fix component name:

```suggestion
The OpenShift Builds operator will introduce a new configuration CRD, `OpenShiftBuild`. It will contain 
```

#### Update domain name for Shared Resource CSI Driver

The root domain/api group used for the Shared Resource CSI driver components will be changed from
`sharedresource.openshift.io` to `sharedresource.dev`, impacting the registered name of the CSI
driver and resource CRDs. To keep the existing feature in OCP, the Shared Resource CSI driver code
may need to be forked to a new GitHub repo/organization.

#### Dependency Resolution with OpenShift Pipelines

Notably, the design of this operator does **_not_** intend to use OLM API dependency resolution
to automate the deployment and management of OpenShift Pipelines. Cluster admins will need to
separately (and manually) install OpenShift Pipelines. This ensures that OpenShift Pipelines is not
installed if a cluster admin does not want Shipwright Builds.

### Risks and Mitigations

#### Duplicate CSI Drivers

On a tech preview cluster, the same CSI driver could be deployed twice on tech preview clusters.
Changing the root domain of the driver and CRDs avoids this problem - the driver and CRDs have
different names/APIGroups, and this are distinct entities in OpenShift. The driver DaemonSet should

> likewise not collide, since it will be deployed in a different namespace.

Author: @gabemontero want to check here that the CSI driver doesn't use any hard-coded paths on the node that would be shared if we had two of these drivers running simultaneously, under different DaemonSets.

Reviewer: good catch @adambkaplan .... there are some hard coded paths on the node with the current OCP driver, so we'll have to make some set of changes with the "operator delivered GA version" ... some combination of
- make different with the new version
- make configurable
- or the new version detects the presence of the old OCP version somehow and adjusts

Reviewer 5: To install the old CSI driver, one needs `TechPreviewNoUpgrade` feature set, which implies no upgrade and the cluster is sort of testing-only. So, do we need to worry about anyone having both CSI drivers?

> The driver DaemonSet should likewise not collide, since it will be deployed in a different namespace.
The CSI driver name (i.e. `csi.sharedresource.openshift.io`) should be different too, just to be sure.

Author: There are two rationales for the rename:

1. Tactical: this gives us a wider install base. We've been promising this thing for a long time, so the more versions we support, the better. I agree that having both CSI drivers is a bit of an edge case.
2. Strategic: we will want to upstream this, either as its own project outright or as something within Kubernetes SIG-Storage. Getting rid of openshift-isms helps with that.

#### Unwanted components/CRDs

Admins or users may want to use the Shared Resource CSI Driver without Shipwright or Tekton - for
example, developers or teams that heavily rely on the `BuildConfig` APIs today. To address this
concern, the OpenShift Builds operator will not rely on OLM API resolution to automate the
installation and deployment of OpenShift Pipelines. The operator will instead report an appropriate

> status message if `spec.shipwright.build.enabled` is set to `true` and it detects that OpenShift

Author: ```suggestion
status message if `spec.shipwright.build.managementState` is set to `Enabled` and it detects that OpenShift
```
